src/dcs/inject.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module hosting the dependency injection container."""
+
+from collections.abc import AsyncGenerator, Coroutine
+from contextlib import asynccontextmanager
+from typing import Callable, Optional
+
+from fastapi import FastAPI
+from ghga_service_commons.api import configure_app
+from ghga_service_commons.auth.jwt_auth import JWTAuthContextProvider
+from hexkit.providers.akafka import KafkaEventPublisher, KafkaEventSubscriber
+from hexkit.providers.mongodb import MongoDbDaoFactory
+from typing_extensions import TypeAlias
+
+from dcs.adapters.inbound.event_sub import EventSubTranslator
+from dcs.adapters.inbound.fastapi_ import dummies
+from dcs.adapters.inbound.fastapi_.custom_openapi import get_openapi_schema
+from dcs.adapters.inbound.fastapi_.routes import router
+from dcs.adapters.outbound.dao import DrsObjectDaoConstructor
+from dcs.adapters.outbound.event_pub import EventPubTranslator
+from dcs.adapters.outbound.s3 import S3ObjectStorage
+from dcs.config import Config
+from dcs.core.auth_policies import WorkOrderContext
+from dcs.core.data_repository import DataRepository
+from dcs.ports.inbound.data_repository import DataRepositoryPort
+from dcs.utils import asyncnullcontext
+
+
+@asynccontextmanager
+async def prepare_core(*, config: Config) -> AsyncGenerator[DataRepositoryPort, None]:
+    """Constructs and initializes all core components and their outbound dependencies.
+    This is a context manager returns a container with all initialized resources upon
+    __enter__. The resources are teared down upon __exit__.
+
+    """
+    dao_factory = MongoDbDaoFactory(config=config)
+    drs_object_dao = await DrsObjectDaoConstructor.construct(dao_factory=dao_factory)
+    object_storage = S3ObjectStorage(config=config)
+
+    async with KafkaEventPublisher.construct(config=config) as event_pub_provider:
+        event_publisher = EventPubTranslator(config=config, provider=event_pub_provider)
+
+        yield DataRepository(
+            drs_object_dao=drs_object_dao,
+            object_storage=object_storage,
+            event_publisher=event_publisher,
+            config=config,
+        )
+
+
+OutboxCleaner: TypeAlias = Callable[[], Coroutine[None, None, None]]
+
+
+def get_configured_rest_app(*, config: Config) -> FastAPI:
```

Comment:
I think this function should be moved to another module, perhaps `utils.py`. This module should strictly be for the injection logic (imo).

Reply:
That is fair.

Reply:
`custom_openapi.py` might be a better place for this function.
Should probably be renamed to something like `app_initialization.py` if we group that together

Reply:
Agree. But I renamed the custom_api module into configure.

Reply:
Please resolve this if you agree.